lib/ansible/modules/cloud/google/gcp_container_cluster.py

```diff
@@ -0,0 +1,925 @@
+
+    if navigate_hash(result, ['error', 'errors']):
+        module.fail_json(msg=navigate_hash(result, ['error', 'errors']))
+
+    return result
+
+
+def is_different(module, response):
+    request = resource_to_request(module)
+    response = response_to_hash(module, response)
+
+    # Remove all output-only from response.
+    response_vals = {}
+    for k, v in response.items():
+        if k in request:
+            response_vals[k] = v
+
+    request_vals = {}
+    for k, v in request.items():
+        if k in response:
+            request_vals[k] = v
+
+    return response_vals != request_vals
+
+
+# Remove unnecessary properties from the response.
+# This is for doing comparisons with Ansible's current parameters.
+def response_to_hash(module, response):
+    return {
+        'name': response.get('name'),
+        'description': response.get('description'),
+        'initial_node_count': response.get('initial_node_count'),
+        'node_config': response.get('node_config'),
+        'master_auth': response.get('master_auth'),
+        'logging_service': response.get('logging_service'),
+        'monitoring_service': response.get('monitoring_service'),
+        'network': response.get('network'),
+        'cluster_ipv4_cidr': response.get('cluster_ipv4_cidr'),
+        'addons_config': response.get('addons_config'),
+        'subnetwork': response.get('subnetwork'),
+        'location': response.get('location'),
+        'endpoint': response.get('endpoint'),
+        'initial_cluster_version': response.get('initial_cluster_version'),
+        'current_master_version': response.get('current_master_version'),
+        'current_node_version': response.get('current_node_version'),
+        'create_time': response.get('create_time'),
+        'node_ipv4_cidr_size': response.get('node_ipv4_cidr_size'),
+        'services_ipv4_cidr': response.get('services_ipv4_cidr'),
+        'current_node_count': response.get('current_node_count'),
+        'expire_time': response.get('expire_time')
+    }
+
+
+def async_op_url(module, extra_data=None):
+    if extra_data is None:
+        extra_data = {}
+    url = "https://container.googleapis.com/v1/projects/{project}/zones/{zone}/operations/{op_id}"
+    combined = extra_data.copy()
+    combined.update(module.params)
+    return url.format(**combined)
+
+
+def wait_for_operation(module, response):
+    op_result = return_if_object(module, response)
+    if op_result is None:
+        return None
+    status = navigate_hash(op_result, ['status'])
+    wait_done = wait_for_completion(status, op_result, module)
+    return fetch_resource(module, navigate_hash(wait_done, ['targetLink']))
+
+
+def wait_for_completion(status, op_result, module):
+    op_id = navigate_hash(op_result, ['name'])
+    op_uri = async_op_url(module, {'op_id': op_id})
+    while status != 'DONE':
+        raise_if_errors(op_result, ['error', 'errors'], 'message')
+        time.sleep(1.0)
```

Comment:
What rate limits are there on this API? Any account-wide ones will cause inconsistent failures if many playbooks are run at once. On the AWS modules we have common exponential-backoff code that handles this, and the base class could be extended for GCP.

Reply:
I haven't been able to find a good answer for the rate limits on this API. I'm not worried about the potential for it to fail though. We can do some kind of exponential-backoff, but I'm thinking that might be better as an additional feature later on